src/dataregistry/query.py

```diff
@@ -0,0 +1,248 @@
+
+def is_orderable_type(ctype):
+    return type(ctype) in ALL_ORDERABLE
+
+class Query():
+    '''
+    Class implementing supported queries
+    '''
+    def __init__(self, db_engine, dialect, schema_version=SCHEMA_VERSION):
+        self._engine = db_engine
+        self._dialect = dialect
+        if dialect == 'sqlite':
+            self._schema_version = None
+        else:
+            self._schema_version=schema_version
+
+        # Do we need to know where the datasets actually are?  If so
+        # we need a ROOT_DIR
+
+        self._metadata_getter = TableMetadata(self._schema_version,
+                                              db_engine)
+
+        # Get table definitions
+        self._tables = dict()
+        self._tables["dataset"] = self._metadata_getter.get("dataset")
+        self._tables["execution"] = self._metadata_getter.get("execution")
+        # self._dataset_alias_table = None
+        # self._execution_alias_table = None
+        # self._dependency_table = None
+        self._dataset_native = None
+        self._from_execution = None
+        self._all_dataset_properties = None
+        self._get_properties()
+
+    def _get_properties(self):
+        '''
+        Cache column names which may be queried.
+        '''
+        if self._dataset_native is None:
+            self._dataset_native = dict()
+            for c in self._tables['dataset'].c:
+                self._dataset_native['dataset.' + c.name] = is_orderable_type(c.type)
+
+        if  self._from_execution is None:
+            self._from_execution = dict()
+            for c in self._tables['execution'].c:
+                self._from_execution['execution.' + c.name] = is_orderable_type(c.type)
+
+    def list_dataset_properties(self):
+        '''
+        Returns a set of property names of the form <table_name>.<column_name>
+        <table_name> is either 'dataset' or another table name (e.g. 'execution')
+        for a table which may be joined with dataset.
+        '''
+        if self._all_dataset_properties:
+            return self._all_dataset_properties.keys()
+
+        all_properties = dict(self._dataset_native)
+        all_properties.update(self._from_execution)
+        self._all_dataset_properties = dict(all_properties)
+        return list(all_properties.keys())
+
+    def is_orderable_property(self, property_name):
+        if not self._all_dataset_properties:
+            list_dataset_properties()
+        if not property_name in self._all_dataset_properties.keys():
+            raise ValueException(f'Unknown property {property_name}')
+        return self._all_dataset_properties[property_name]
+
+    # Do we need this routine?  Will users want to query executions?
+    def list_execution_properties(self):
+        '''
+        Should return a list of property names, not necessarily identical
+        to column names in the execution table.
+        '''
+        pass
```

Comment:
could have `raise NotImplementedError` instead for these, so that we get an error when calling them (which we shouldnt be before they are actually doing something?)

Reply:
I'm not sure we need to implement it at all; it depends on whether we expect people to look up properties of executions explicitly, rather than as part of a query on datasets.  But I agree either it should raise NotImplementedError or not be there at all.